src/integrationTest/sam.test.ts

```diff
@@ -0,0 +1,192 @@
+    { name: 'dotnetcore2.1', path: 'testProject/src/HelloWorld/Function.cs', debuggerType: 'coreclr' }
+]
+
+async function openSamProject(): Promise<vscode.Uri> {
+    const documentPath = path.join(projectFolder, projectPath)
+    await vscode.workspace.openTextDocument(documentPath)
+
+    return vscode.Uri.file(documentPath)
+}
+
+function tryRemoveProjectFolder() {
+    try {
+        removeSync(path.join(projectFolder, 'testProject'))
+    } catch (e) {}
+}
+
+async function getCodeLenses(): Promise<vscode.CodeLens[]> {
+    let codeLenses: vscode.CodeLens[] | undefined
+    while (true) {
+        try {
+            // this works without a sleep locally, but not on CodeBuild
+            await sleep(200)
+            const codeLensesPromise: Thenable<vscode.CodeLens[] | undefined> = vscode.commands.executeCommand(
+                'vscode.executeCodeLensProvider',
+                documentUri
+            )
+            codeLenses = await codeLensesPromise
+            if (!codeLenses) {
+                continue
+            }
+            // omnisharp spits out some undefined code lenses for some reason, we filter them because they are
+            // not shown to the user and do not affect how our extension is working
+            codeLenses = codeLenses.filter(lens => lens !== undefined && lens.command !== undefined)
+            if (codeLenses.length === 3) {
+                return codeLenses as vscode.CodeLens[]
+            }
+        } catch (e) {}
+    }
+}
+
+async function getCodeLensesOrTimeout(): Promise<vscode.CodeLens[]> {
+    const codeLensPromise = getCodeLenses()
+    const timeout = new Promise(resolve => {
+        setTimeout(resolve, 10000, undefined)
+    })
+    const result = await Promise.race([codeLensPromise, timeout])
+
+    if (result) {
+        return result as vscode.CodeLens[]
+    }
+    throw new Error('Codelenses took too long to show up, this inidicates an issue!')
+}
+
+async function onDebugChanged(e: vscode.DebugSession | undefined) {
+    if (!e) {
+        return
+    }
+    assert.strictEqual(e.configuration.name, 'SamLocalDebug')
+    assert.strictEqual(e.configuration.type, debuggerType)
+    // wait for it to actually start (which we do not get an event for). 800 is
+    // short enough to finish before the next test is run and long enough to
+    // actually act after it pauses
+    await sleep(800)
+    await vscode.commands.executeCommand('workbench.action.debug.continue')
+}
+
+// Iterate through and test all runtimes
+for (const runtime of runtimes) {
+    describe(`SAM Integration tests ${runtime.name}`, async () => {
+        before(async function() {
+            // tslint:disable-next-line: no-invalid-this
+            this.timeout(TIMEOUT)
+            projectSDK = runtime.name
+            projectPath = runtime.path
+            debuggerType = runtime.debuggerType
+            // set up debug config
+            vscode.debug.onDidChangeActiveDebugSession(onDebugChanged)
```

Comment:
Any event that is registered returns a disposable. The disposable needs to be disposed at the end of each test (after or afterEach, depending on the situation) or it will continue to handle events.

Reply:
sure added